libnmstate/dns.py

```diff
@@ -0,0 +1,205 @@
+#
+# Copyright (c) 2020 Red Hat, Inc.
+#
+# This file is part of nmstate
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+from copy import deepcopy
+
+from libnmstate.error import NmstateValueError
+from libnmstate.error import NmstateVerificationError
+from libnmstate.error import NmstateNotImplementedError
+from libnmstate.iplib import is_ipv6_address
+from libnmstate.prettystate import format_desired_current_state_diff
+from libnmstate.schema import DNS
+from libnmstate.schema import Interface
+
+
+class DnsState:
+    DNS_METADATA = "_dns"
+    PRIORITY_METADATA = "_priority"
+
+    def __init__(self, ifaces, routes, des_dns_state, cur_dns_state):
+        self._ifaces = ifaces
+        self._routes = routes
+        self._config_changed = False
+        self._cur_dns_state = deepcopy(cur_dns_state) if cur_dns_state else {}
+        if des_dns_state:
+            self._dns_state = des_dns_state
+            if self._config_servers != _get_config_servers(
+                cur_dns_state
+            ) or self._config_searches != _get_config_searches(cur_dns_state):
+                self._config_changed = True
+            self._validate_dns()
+        else:
+            self._dns_state = cur_dns_state if cur_dns_state else {}
+        if ifaces and routes:
+            self._gen_metadata()
+
+    @property
+    def current_config(self):
+        return _get_config(self._cur_dns_state)
+
+    @property
+    def config(self):
+        return _get_config(self._dns_state)
+
+    @property
+    def _config_servers(self):
+        return _get_config_servers(self._dns_state)
+
+    @property
+    def _config_searches(self):
+        return _get_config_searches(self._dns_state)
+
+    def _gen_metadata(self):
+        """
+        Store DNS configurations to interface as metadata
```

Comment:
So you mutate the interface state with DNS metadata in the DNS object?
It feels dangerous to me.

I would have preferred to see metadata info being prepared here and someone else (the iface object possibly) mutating itself. Self mutation is preferred over mutating others, the later is unsafe and not under the control of the object that represents that data.

It is also interesting to note that only the metadata generation is causing the dns class to be dependent on the routes and ifaces states, which means that metadata is worth extracting out to its own class that gets as input a dns, iface and route object and composes from all the needed (avoiding the dependencies between them, moving the dependency outside).

Reply:
Good point.
I have changed to work flow to:
 * NetState instruct `Ifaces` to `gen_dns_metadata()`.
 * Then `DnsState.gen_metadata()` return a structured data which could be used by `Ifaces` to store metatadata in the proper way.
 * The `BaseIface.store_dns_metadata()` will store it.

Similar approach for route and route rule.